spec/models/data_file_spec.rb

```diff
@@ -262,53 +262,100 @@
 
-      context 'being deleted' do
-        subject { child_file }
-        it {
-          expect(subject.parent_id).not_to be_nil
-          expect(subject.parent).not_to be_nil
-          original_parent = subject.parent
-          expect(subject.deleted_from_parent_id).to be_nil
-          expect(subject.deleted_from_parent).to be_nil
-
-          subject.is_deleted = true
-          subject.set_deleted_from_parent
-
-          expect(subject.parent_id).to be_nil
-          expect(subject.parent).to be_nil
-          expect(subject.deleted_from_parent_id).not_to be_nil
-          expect(subject.deleted_from_parent).not_to be_nil
-          expect(subject.deleted_from_parent).to eq(original_parent)
-        }
-      end
+        subject.move_to_trashbin
+
+        expect(subject.is_deleted?).to be_truthy
+        expect(subject.parent_id).to be_nil
+        expect(subject.parent).to be_nil
+        expect(subject.deleted_from_parent_id).not_to be_nil
+        expect(subject.deleted_from_parent).not_to be_nil
+        expect(subject.deleted_from_parent).to eq(original_parent)
+      }
+    end
+  end
+
+  describe '.restore_from_trashbin' do
+    it { is_expected.to respond_to(:restore_from_trashbin) }
+
+    context 'to original parent' do
+      it {
+        subject.move_to_trashbin
+        subject.save
+
+        expect(subject.is_deleted?).to be_truthy
+        expect(subject.parent_id).to be_nil
+        expect(subject.parent).to be_nil
+        expect(subject.deleted_from_parent_id).not_to be_nil
+        expect(subject.deleted_from_parent).not_to be_nil
+        original_parent = subject.deleted_from_parent
+
+        subject.restore_from_trashbin
+
+        expect(subject.is_deleted?).to be_falsey
+        expect(subject.parent_id).not_to be_nil
+        expect(subject.parent).not_to be_nil
+        expect(subject.deleted_from_parent_id).to be_nil
+        expect(subject.deleted_from_parent).to be_nil
+        expect(subject.parent).to eq(original_parent)
+      }
+    end
+
+    context 'to new parent folder' do
+      let(:new_parent) { other_folder }
+      it {
+        subject.move_to_trashbin
+        subject.save
+
+        expect(subject.is_deleted?).to be_truthy
+        expect(subject.parent_id).to be_nil
+        expect(subject.parent).to be_nil
+        expect(subject.deleted_from_parent_id).not_to be_nil
+        expect(subject.deleted_from_parent).not_to be_nil
+        original_parent = subject.deleted_from_parent
+
+        subject.restore_from_trashbin(other_folder)
+
+        expect(subject.is_deleted?).to be_falsey
+        expect(subject.parent_id).not_to be_nil
+        expect(subject.parent).not_to be_nil
+        expect(subject.deleted_from_parent_id).to be_nil
+        expect(subject.deleted_from_parent).to be_nil
+        expect(subject.parent).not_to eq(original_parent)
+        expect(subject.parent).to eq(new_parent)
+      }
+    end
+
+    context 'to project root' do
```

Comment:
This tests how the method behaves when passed the subject's project. What happens with a different project?

Reply:
I'm also curious what happens when it's passes a folder from a different project.

Reply:
The function does what it is told to do when a new project, or folder from a new project is passed. It sets these variables. This results in an invalid object based on the validations (project is immutable), which the trashbin_api restore method now communicates to the user.